pennylane/optimize/rotosolve.py

```diff
@@ -14,78 +14,122 @@
+    elif optimizer == "shgo":
+        optimizer = _shgo_optimizer
+        optimizer_kwargs.setdefault("bounds", ((-np.pi, np.pi),))
+
+    return optimizer, optimizer_kwargs
+
+
+def _restrict_to_univariate(fn, arg_idx, par_idx, args, kwargs):
+    r"""Restrict a function to a univariate function for given argument
+    and parameter indices.
+
+    Args:
+        fn (callable): Multivariate function
+        arg_idx (int): Index of the argument that contains the parameter to restrict to
+        par_idx (tuple[int]): Index of the parameter to restrict to within the argument
+        args (tuple): Arguments at which to restrict the function.
+        kwargs (dict): Keyword arguments at which to restrict the function.
+
+    Returns:
+        callable: Univariate restriction of ``fn``. That is, this callable takes
+        a single float value as input and has the same return type as ``fn``.
+        All arguments are set to the given ``args`` and the input value to this
+        function is added to the marked parameter.
+    """
+    the_arg = args[arg_idx]
+    if len(qml.math.shape(the_arg)) == 0:
+        shift_vec = qml.math.ones_like(the_arg)
+    else:
+        shift_vec = qml.math.zeros_like(the_arg)
+        shift_vec = qml.math.scatter_element_add(shift_vec, par_idx, 1.0)
+
+    def _univariate_fn(x):
+        return fn(*args[:arg_idx], the_arg + shift_vec * x, *args[arg_idx + 1 :], **kwargs)
+
+    return _univariate_fn
+
+
+def _min_analytic(objective_fn, freq, f0):
+    r"""Analytically minimize a trigonometric function that depends on a
+    single parameter and has a single frequency. Uses two or
+    three function evaluations.
+
+    Args:
+        objective_fn (callable): Trigonometric function to minimize
+        freq (float): Frequency in the ``objective_fn``
+        f0 (float): Value of the ``objective_fn`` at zero. Reduces the
+            number of calls to the function from three to two if given.
+
+    Returns:
+        float: Position of the minimum of ``objective_fn``
+        float: Value of the minimum of ``objective_fn``
+
+    The closed form expression used here was derived in
+    `Vidal & Theis (2018) <https://arxiv.org/abs/1812.06323>`__ ,
+    `Parrish et al (2019) <https://arxiv.org/abs/1904.03206>`__ and
+    `Ostaszewski et al (2019) <https://arxiv.org/abs/1905.09692>`__.
+    We use the notation of Appendix A of the latter reference, allowing
+    for an arbitrary frequency instead of restricting to ``freq=1``.
+    The returned position is guaranteed to lie within :math:`(-\pi, \pi]`.
+    """
+    if f0 is None:
+        f0 = objective_fn(0.0)
+    fp = objective_fn(0.5 * np.pi / freq)
+    fm = objective_fn(-0.5 * np.pi / freq)
+    C = 0.5 * (fp + fm)
+    B = np.arctan2(2 * f0 - fp - fm, fp - fm)
+    x_min = (-np.pi / 2 - B) / freq
+    A = np.sqrt((f0 - C) ** 2 + 0.25 * (fp - fm) ** 2)
+    y_min = -A + C
+
+    if x_min <= -np.pi:
+        x_min = x_min + 2 * np.pi
+
+    return x_min, y_min
```

Comment:
This is very nice! I wonder if it is worth having as a public static method on `RotosolveOptimizer, just so that it is documented?

This is very minor though, so feel free to resolve.

Reply:
I like the idea! :)

Reply:
Included it as a `staticmethod`